Creating EmbeddingModel failed

info:
when creating the embeddingModel, running it on the Mac was successful, but deploying it online on Linux reported an error

errorMsg:
`Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [dev.langchain4j.model.embedding.EmbeddingModel]: Factory method 'embeddingModel' threw exception with message: /home/admin/upper/.default/temp/onnxruntime-java3822066698664609290/libonnxruntime.so: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by /home/admin/upper/.default/temp/onnxruntime-java3822066698664609290/libonnxruntime.so)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:177)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:651)
	... 54 more
Caused by: java.lang.UnsatisfiedLinkError: /home/admin/upper/.default/temp/onnxruntime-java3822066698664609290/libonnxruntime.so: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by /home/admin/upper/.default/temp/onnxruntime-java3822066698664609290/libonnxruntime.so)
	at java.base/jdk.internal.loader.NativeLibraries.load(Native Method)
	at java.base/jdk.internal.loader.NativeLibraries$NativeLibraryImpl.open(NativeLibraries.java:331)
	at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:197)
	at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:139)
	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2418)
	at java.base/java.lang.Runtime.load0(Runtime.java:852)
	at java.base/java.lang.System.load(System.java:2025)
	at ai.onnxruntime.OnnxRuntime.load(OnnxRuntime.java:365)
	at ai.onnxruntime.OnnxRuntime.init(OnnxRuntime.java:156)
	at ai.onnxruntime.OrtEnvironment.<clinit>(OrtEnvironment.java:33)
	at dev.langchain4j.model.embedding.OnnxBertBiEncoder.<init>(OnnxBertBiEncoder.java:36)
	at dev.langchain4j.model.embedding.AbstractInProcessEmbeddingModel.loadFromJar(AbstractInProcessEmbeddingModel.java:22)
	at dev.langchain4j.model.embedding.AllMiniLmL6V2EmbeddingModel.<clinit>(AllMiniLmL6V2EmbeddingModel.java:18)
	at com.alibaba.tre.configuration.LLMConfiguration.embeddingModel(LLMConfiguration.java:29)
	at com.alibaba.tre.configuration.LLMConfiguration$$SpringCGLIB$$0.CGLIB$embeddingModel$2(<generated>)
	at com.alibaba.tre.configuration.LLMConfiguration$$SpringCGLIB$$FastClass$$1.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:258)
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331)
	at com.alibaba.tre.configuration.LLMConfiguration$$SpringCGLIB$$0.embeddingModel(<generated>)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:140)
	... 55 more`

--- a/langchain4j-dashscope/pom.xml
+++ b/langchain4j-dashscope/pom.xml
@@ -24,7 +24,7 @@
         <dependency>
             <groupId>com.alibaba</groupId>
             <artifactId>dashscope-sdk-java</artifactId>
-            <version>2.14.7</version>
+            <version>2.15.1</version>
         </dependency>
 
         <dependency>

--- a/langchain4j-dashscope/src/main/java/dev/langchain4j/model/dashscope/QwenHelper.java
+++ b/langchain4j-dashscope/src/main/java/dev/langchain4j/model/dashscope/QwenHelper.java
@@ -7,6 +7,7 @@ import com.alibaba.dashscope.aigc.multimodalconversation.MultiModalConversationO
 import com.alibaba.dashscope.aigc.multimodalconversation.MultiModalConversationResult;
 import com.alibaba.dashscope.common.Message;
 import com.alibaba.dashscope.common.MultiModalMessage;
+import com.alibaba.dashscope.common.Role;
 import com.alibaba.dashscope.tools.*;
 import com.alibaba.dashscope.utils.JsonUtils;
 import dev.langchain4j.agent.tool.ToolExecutionRequest;
@@ -18,6 +19,7 @@ import dev.langchain4j.data.message.*;
 import dev.langchain4j.internal.Utils;
 import dev.langchain4j.model.output.FinishReason;
 import dev.langchain4j.model.output.TokenUsage;
+import lombok.extern.slf4j.Slf4j;
 
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
@@ -26,18 +28,22 @@ import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.nio.file.StandardCopyOption;
 import java.util.*;
+import java.util.function.BiFunction;
+import java.util.function.BinaryOperator;
 import java.util.stream.Collectors;
 
-import static com.alibaba.dashscope.common.Role.*;
+import static dev.langchain4j.data.message.ChatMessageType.*;
 import static dev.langchain4j.internal.Utils.getOrDefault;
 import static dev.langchain4j.internal.Utils.isNullOrEmpty;
 import static dev.langchain4j.model.output.FinishReason.*;
 import static java.util.stream.Collectors.toList;
 
+@Slf4j
 class QwenHelper {
 
     static List<Message> toQwenMessages(List<ChatMessage> messages) {
-        return messages.stream()
+        return sanitizeMessages(messages)
+                .stream()
                 .map(QwenHelper::toQwenMessage)
                 .collect(toList());
     }
@@ -97,7 +103,7 @@ class QwenHelper {
     }
 
     static List<ToolCallBase> toolCallsFrom(ChatMessage message) {
-        if (message.type() == ChatMessageType.AI && ((AiMessage) message).hasToolExecutionRequests()) {
+        if (message.type() == AI && ((AiMessage) message).hasToolExecutionRequests()) {
             return toToolCalls(((AiMessage) message).toolExecutionRequests());
         }
         return null;
@@ -189,14 +195,14 @@ class QwenHelper {
     }
 
     static String roleFrom(ChatMessage message) {
-        if (message.type() == ChatMessageType.AI) {
-            return ASSISTANT.getValue();
-        } else if (message.type() == ChatMessageType.SYSTEM) {
-            return SYSTEM.getValue();
+        if (message.type() == AI) {
+            return Role.ASSISTANT.getValue();
+        } else if (message.type() == SYSTEM) {
+            return Role.SYSTEM.getValue();
         } else if (message.type() == ChatMessageType.TOOL_EXECUTION_RESULT) {
-            return TOOL.getValue();
+            return Role.TOOL.getValue();
         } else {
-            return USER.getValue();
+            return Role.USER.getValue();
         }
     }
 
@@ -401,5 +407,68 @@ class QwenHelper {
         callFunction.setArguments(toolExecutionRequest.arguments());
         toolCallFunction.setFunction(callFunction);
         return toolCallFunction;
+    }
+
+    static List<ChatMessage> sanitizeMessages(List<ChatMessage> messages) {
+        LinkedList<ChatMessage> sanitizedMessages = messages.stream()
+                .reduce(new LinkedList<>(), messageAccumulator(), messageCombiner());
+
+        // Ensure the last message is a user/tool_execution_result message
+        while(!sanitizedMessages.isEmpty() && !isInputMessageType(sanitizedMessages.getLast().type())) {
+            ChatMessage removedMessage = sanitizedMessages.removeLast();
+            log.warn("The last message should be a user/tool_execution_result message, but found: {}", removedMessage);
+        }
+
+        return sanitizedMessages;
+    }
+
+    private static BiFunction<LinkedList<ChatMessage>, ChatMessage, LinkedList<ChatMessage>> messageAccumulator() {
+        return (acc, message) -> {
+            ChatMessageType type = message.type();
+            if (acc.isEmpty()) {
+                // Ensure the first message is a system message or a user message.
+                if (type == SYSTEM || type == USER) {
+                    acc.add(message);
+                } else {
+                    log.warn("The first message should be a system message or a user message, but found: {}", message);
+                }
+                return acc;
+            }
+
+            if (type == SYSTEM) {
+                // Ensure the system message is the first message.
+                log.warn("The system message should be the first message. Drop existed messages: {}", acc);
+                acc.clear();
+                acc.add(message);
+                return acc;
+            }
+
+            ChatMessageType lastType = acc.getLast().type();
+            if (lastType == SYSTEM && type != USER) {
+                // The first non-system message must be a user message.
+                log.warn("The first non-system message must be a user message, but found: {}", message);
+                return acc;
+            }
+
+            if (isInputMessageType(type) == isInputMessageType(lastType)) {
+                // The list must be user/tool_execution_result and ai alternating messages.
+                // Use the newest one when duplicated.
+                ChatMessage removedMessage = acc.removeLast();
+                log.warn("User/Tool-execution-result messages and AI messages should alternate. Drop duplicated message: {}", removedMessage);
+            }
+
+            acc.add(message);
+            return acc;
+        };
+    }
+
+    private static BinaryOperator<LinkedList<ChatMessage>> messageCombiner() {
+        return (acc1, acc2) -> {
+            throw new UnsupportedOperationException("Parallel stream not supported");
+        };
+    }
+
+    private static boolean isInputMessageType(ChatMessageType messageType) {
+        return messageType == USER || messageType == TOOL_EXECUTION_RESULT;
     }
 }

--- a/langchain4j-dashscope/src/test/java/dev/langchain4j/model/dashscope/QwenChatModelIT.java
+++ b/langchain4j-dashscope/src/test/java/dev/langchain4j/model/dashscope/QwenChatModelIT.java
@@ -3,17 +3,16 @@ package dev.langchain4j.model.dashscope;
 import dev.langchain4j.agent.tool.JsonSchemaProperty;
 import dev.langchain4j.agent.tool.ToolExecutionRequest;
 import dev.langchain4j.agent.tool.ToolSpecification;
-import dev.langchain4j.data.message.AiMessage;
-import dev.langchain4j.data.message.ChatMessage;
-import dev.langchain4j.data.message.ToolExecutionResultMessage;
-import dev.langchain4j.data.message.UserMessage;
+import dev.langchain4j.data.message.*;
 import dev.langchain4j.model.chat.ChatLanguageModel;
 import dev.langchain4j.model.output.Response;
 import dev.langchain4j.model.output.TokenUsage;
+import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.MethodSource;
 
+import java.util.LinkedList;
 import java.util.List;
 
 import static dev.langchain4j.agent.tool.JsonSchemaProperty.INTEGER;
@@ -235,5 +234,67 @@ public class QwenChatModelIT {
         System.out.println(response);
 
         assertThat(response.content().text()).containsIgnoringCase("parrot");
+    }
+
+    @Test
+    public void should_sanitize_messages() {
+        List<ChatMessage> messages = new LinkedList<>();
+
+        // 1. The system message should be the first message.
+        // 2. User/Tool-execution-result messages and AI messages should alternate.
+        // 3. The last message in the message list should be a user message. This serves as the model query/input for the current round.
+
+        messages.add(SystemMessage.from("System message 1, which should be discarded"));
+        messages.add(UserMessage.from("User message 1, which should be discarded"));
+        messages.add(SystemMessage.from("System message 2"));
+
+        messages.add(AiMessage.from("AI message 1, which should be discarded"));
+        messages.add(ToolExecutionResultMessage.from(ToolExecutionRequest.builder().build(),
+                "Tool execution result 1, which should be discards"));
+        messages.add(UserMessage.from("User message 2, which should be discarded"));
+        messages.add(UserMessage.from("User message 3"));
+
+        messages.add(AiMessage.from("AI message 2, which should be discarded"));
+        messages.add(AiMessage.from("AI message 3"));
+
+        messages.add(ToolExecutionResultMessage.from(ToolExecutionRequest.builder().build(),
+                "Tool execution result 2, which should be discards"));
+        messages.add(ToolExecutionResultMessage.from(ToolExecutionRequest.builder().build(),
+                "Tool execution result 3"));
+
+        messages.add(AiMessage.from("AI message 4"));
+
+        messages.add(UserMessage.from("User message 4, which should be discards"));
+        messages.add(UserMessage.from("User message 5"));
+
+        messages.add(AiMessage.from("AI message 5, which should be discards"));
+
+        // The result should be in the following order:
+        // 1. System message
+        // 2. User message
+        // 3. AI message
+        // 4. Tool execution result message
+        // 5. AI message
+        // 6. User message
+        List<ChatMessage> sanitizedMessages = QwenHelper.sanitizeMessages(messages);
+        assertThat(sanitizedMessages).hasSize(6);
+
+        assertThat(sanitizedMessages.get(0)).isInstanceOf(SystemMessage.class);
+        assertThat(((SystemMessage) sanitizedMessages.get(0)).text()).isEqualTo("System message 2");
+
+        assertThat(sanitizedMessages.get(1)).isInstanceOf(UserMessage.class);
+        assertThat(((UserMessage) sanitizedMessages.get(1)).singleText()).isEqualTo("User message 3");
+
+        assertThat(sanitizedMessages.get(2)).isInstanceOf(AiMessage.class);
+        assertThat(((AiMessage) sanitizedMessages.get(2)).text()).isEqualTo("AI message 3");
+
+        assertThat(sanitizedMessages.get(3)).isInstanceOf(ToolExecutionResultMessage.class);
+        assertThat(((ToolExecutionResultMessage) sanitizedMessages.get(3)).text()).isEqualTo("Tool execution result 3");
+
+        assertThat(sanitizedMessages.get(4)).isInstanceOf(AiMessage.class);
+        assertThat(((AiMessage) sanitizedMessages.get(4)).text()).isEqualTo("AI message 4");
+
+        assertThat(sanitizedMessages.get(5)).isInstanceOf(UserMessage.class);
+        assertThat(((UserMessage) sanitizedMessages.get(5)).singleText()).isEqualTo("User message 5");
     }
 }

